File: apps/desktop/src/api/rawMaterials.ts
import { invoke } from "@tauri-apps/api/core";
import type { RawMaterial, RawMaterialInput } from "./types";

/**
 * Typed wrappers around the `commands::raw_materials` Tauri commands. This is the only module
 * that should call `invoke()` for raw materials — components call these functions instead.
 */

export function listRawMaterials(includeInactive: boolean): Promise<RawMaterial[]> {
  return invoke("list_raw_materials", { includeInactive });
}

export function getRawMaterial(id: number): Promise<RawMaterial> {
  return invoke("get_raw_material", { id });
}

export function createRawMaterial(input: RawMaterialInput): Promise<RawMaterial> {
  return invoke("create_raw_material", { input });
}

export function updateRawMaterial(id: number, input: RawMaterialInput): Promise<RawMaterial> {
  return invoke("update_raw_material", { id, input });
}

export function archiveRawMaterial(id: number): Promise<void> {
  return invoke("archive_raw_material", { id });
}

export function reactivateRawMaterial(id: number): Promise<void> {
  return invoke("reactivate_raw_material", { id });
}

export function deleteRawMaterial(id: number): Promise<void> {
  return invoke("delete_raw_material", { id });
}

File: apps/desktop/src-tauri/src/commands/raw_materials.rs
use tauri::State;

use crate::db::repositories::raw_materials::{self, RawMaterial, RawMaterialInput};
use crate::error::{AppError, AppResult};
use crate::DbState;

#[tauri::command]
pub fn list_raw_materials(
    db: State<DbState>,
    include_inactive: bool,
) -> AppResult<Vec<RawMaterial>> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::list(&conn, include_inactive)
}

#[tauri::command]
pub fn get_raw_material(db: State<DbState>, id: i64) -> AppResult<RawMaterial> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::get(&conn, id)
}

#[tauri::command]
pub fn create_raw_material(db: State<DbState>, input: RawMaterialInput) -> AppResult<RawMaterial> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::create(&conn, input)
}

#[tauri::command]
pub fn update_raw_material(
    db: State<DbState>,
    id: i64,
    input: RawMaterialInput,
) -> AppResult<RawMaterial> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::update(&conn, id, input)
}

#[tauri::command]
pub fn archive_raw_material(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::set_active(&conn, id, false)
}

#[tauri::command]
pub fn reactivate_raw_material(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::set_active(&conn, id, true)
}

#[tauri::command]
pub fn delete_raw_material(db: State<DbState>, id: i64) -> AppResult<()> {
    let conn = db.0.lock().map_err(|e| AppError::new(e.to_string()))?;
    raw_materials::delete(&conn, id)
}
